daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/SpecifiedLengthUnparsers.scala

```diff
@@ -428,3 +429,20 @@ class SpecifiedLengthPrefixedUnparser(
   }
 
 }
+
+class ChoiceLengthUnparser(
+  eUnparser: Unparser,
+  rd: TermRuntimeData)
+  extends CombinatorUnparser(rd)
+  with CaptureUnparsingValueLength {
+
+  override lazy val runtimeDependencies = Vector()
+  override lazy val childProcessors = Vector(eUnparser)
+
+  override def unparse(state: UState): Unit = {
+    val elem = state.currentInfosetNode.asInstanceOf[DIElement]
+    captureValueLengthStart(state, elem)
+    eUnparser.unparse1(state)
+    captureValueLengthEnd(state, elem)
+  }
```

Comment:
I think this is where things would need to change to create a new thing in the infoset where we can hang length information. Note that I think in the example I gave above, this would capture different lengths on the "foo" elem when each choice unparse was run.

Reply:
I guess I don't see the need for something on the infoset. Choices have no corresponding structure in the infoset. Can you motivate this idea further?

Reply:
Actually, this is harder than that.
Due to the fact that the contents of the choice may have interior alignment, or variable-length things in it computed by dfdl:outputValueCalc, we can't just unparse, and know the length at the point the call returns.

I have almost the identical problem in the new separators code. I need to measure how big something is when unparsing, and decide to lay down a separator or not depending on whether it came out to be zero-length or not. 

Here the situation is almost the same I think. We need to unparse, measure relative to the choiceLength, and skip the unused bytes. But we can't know that length until, well, it's known. 

So the technique is to unparse a suspendable unparser that does nothing, just causes a new DataOutputStream to be split off of the prior, so as to establish a base position. Then unparse the choice, then unparse a suspendable unparser that does a skip N, where N is computed based on the starting position and ending positions of the data output streams being known, and the choiceLength. This would also be able to signal an error if it is too long. 

If that makes sense to everyone, then I suggest we finish review on the separators/sequences branch, merge it, and use the suspendable unparsers created there (one of them - the does-nothing one - can be reused).

Reply:
So in the last commit I pushed up I resolved all the other issues and rebased onto the separator suppression code.  I think I know what needs to be done, but want to make sure I am understanding everything correctly.  Should I be changing ChoiceLengthUnparser to extend it with SuspendableUnparser, and then setting the suspendableOperation to SuppressionRegionSplitSuspendableOperation?  If so, should I move SuppressionRegionSplitSuspendableOperation out into its some sort of common area, as my use here has nothing to do with suppressing separators.

Reply:
You probably want ChoiceLengthUnparser to recursively unparse 

1) a splitter unparser - like in the separator thing, you just need to split the infoset. (Yes this could be hoisted out to a more general class name, since it really has nothing to do with separators per se.)

2) the choice branch itself, and then 

3) a suspendable unparser for the ChoiceUnusedRegion that comes after the choice branch. This will be a new suspendable unparser. The constructor for this unparser will take the DOS from after the split(1) above as an argument. 

The test for the suspendable operation of this ChoiceUnusedRegion unparser is going to use the dataOutputStream from the split (1) above. The test will pass when that data stream, and all subsequent ones up until but not including the current one (the ChoiceUnusedRegion is suspendable, so will also create a split of its own) are final. The current DOS is waiting for us to put the skip(n) into it for the unused region, so this one will NOT be final. 

The continuation method of the suspendable operations can then compute the exact length in bits of the data from the split (1) to just before the ChoiceUnusedRegion unparser (3) by following the chain of DOS from the first one created after the split (1) to current. That's the key value you need to then decide how many bits to skip based on the choiceLength value. 
You skip(n) those, then finalize the DOS. 

Hope that makes sense. Happy to discuss by phone.

Reply:
I believe I have now implemented this correctly in the most recent commit

Reply:
One thing to keep in mind. The ChoiceLengthUnparser has slots in it that capture state. That means these must be created on the fly as needed and are not something that can be serialized as part of a saved parser/unparser, because those are shared state that can be used by multiple threads simultaneously. These ChoiceLengthUnparser instances must be per-thread.

Reply:
Instead of doing something like thread locals, can the capture/splitter logic just be moved out of those functions and into def unparse? And then the maybeDOStart/End vars just become local vals in def unparse that get passed into the creation of the ChoiceUnusedUnparser?

Reply:
They have to live on the heap, they outlive the call to the unparse method. But they can live in a heap allocated object that is part of the suspension state of the unparse invocation.